ref.c

```diff
@@ -0,0 +1,406 @@
+    int local_path = 0;
+
+    if (!ref_path || *ref_path == '\0') {
+        /*
+        * If we have no ref path, we use the EBI server.
+        * However to avoid spamming it we require a local ref cache too.
+        */
+
+        ref_path = "http://www.ebi.ac.uk:80/ena/cram/md5/%s";
+        if (!local_cache || *local_cache == '\0') {
+            const char *extra;
+            const char *base = get_cache_basedir(&extra);
+            snprintf(cache_root, PATH_MAX, "%s%s/hts-ref", base, extra);
+            snprintf(cache, PATH_MAX, "%s%s/hts-ref/%%2s/%%2s/%%s", base, extra);
+            local_cache = cache;
+            hts_log_info("Populating local cache: %s", local_cache);
+        }
+    }
+
+    /* Use cache if available */
+    if (local_cache && *local_cache) {
+        expand_cache_path(path, local_cache, (char *)m5_str);
+        local_path = 1;
+    }
+
+#ifndef HAVE_MMAP
+    char *path2;
+    /* Search local files in REF_PATH; we can open them and return as above */
+    if (!local_path && (path2 = find_path((char *)m5_str, ref_path))) {
+        strncpy(path, path2, PATH_MAX);
+        free(path2);
+        if (is_file(path)) // incase it's too long
+            local_path = 1;
+    }
+#endif
+
+    if (local_path){
+        struct stat sb;
+        hFILE* innerhf;
+
+        if((0 == stat(path, &sb)) && (innerhf = hopen(path, "r"))){
+            /* Found via REF_CACHE or local REF_PATH file */
+            ref = open_hfile_ref(innerhf, path, sb.st_size, NULL, 0);
+
+            return ref;
+        }
+    }
+
+    mFILE* mf;
+    /* Look in ref_path */
+    if (!(mf = open_path_mfile((char*)m5_str, ref_path, NULL))) {
+        hts_log_info("Failed to fetch file. REF_PATH: '%s', M5: '%s'", ref_path, m5_str);
+        return NULL;
+    }
+
+    ref = mFILE_to_hFILE_ref(mf);
+
+    /* Populate the local disk cache if required */
+    if (local_cache && *local_cache) {
+        int pid = (int)getpid();
+        unsigned thrid = get_int_threadid();
+        hFILE *fp;
+
+        if (*cache_root && !is_directory(cache_root)) {
+            hts_log_warning("Creating reference cache directory %s\n"
+                            "This may become large; see the samtools(1) manual page REF_CACHE discussion",
+                            cache_root);
+        }
+
+        expand_cache_path(path, local_cache, (char *)m5_str);
+        hts_log_info("Writing cache file '%s'", path);
+        mkdir_prefix(path, 01777);
+
+        do {
+            // Attempt to further uniquify the temporary filename
+            unsigned t = ((unsigned)time(NULL)) ^ ((unsigned)clock());
+            thrid++; // Ensure filename changes even if time/clock haven't
+
+            sprintf(path_tmp, "%s.tmp_%d_%u_%u", path, pid, thrid, t);
+            fp = hopen(path_tmp, "wx");
+        } while (fp == NULL && errno == EEXIST);
+        if (!fp) {
+            perror(path_tmp);
+
+            // Not fatal - we have the data already so keep going.
+        }
```

Comment:
We lost a `return 0` (NULL now) in the `if (!fp)` clause from the original cram_io.c copy.  The intent was if we fail to create a local temporary file name to cache the reference, then so be it - we do still have our reference and the code can continue to work, but for whatever reason (eg permission denied) being able to write to the cache is a non-fatal condition.

However continuing means we end up calling hwrite, hfclose etc on NULL.

Reply:
I have validated that this is an issue and that adding `return 0` after the comment fixes it.
Try setting REF_CACHE to a directory that you don't have write access to and running the test-ref tool.  It'll segmentation fault.